torch_xla/csrc/aten_xla_type.cpp

```diff
@@ -359,6 +359,34 @@ at::Tensor AtenXlaType::_copy_from(const at::Tensor& self,
   return dst;
 }
 
+at::Tensor AtenXlaType::_copy_from_and_resize(const at::Tensor& self,
+                                              const at::Tensor& dst) {
+  XLA_FN_COUNTER("xla::");
+  auto dst_tensor = bridge::TryGetXlaTensor(dst);
+  auto self_tensor = bridge::TryGetXlaTensor(self);
+  if (!self_tensor) {
+    XLA_CHECK(dst_tensor);
+    dst_tensor->UpdateFromTensorOut(self);
+  } else if (!dst_tensor) {
+    at::Tensor tensor = self_tensor->ToTensor(/*detached=*/true);
+    at::Tensor typed_tensor =
+        CopyTensor(tensor, dst.scalar_type(), /*copy=*/false);
+    dst.resize_as_(typed_tensor).copy_(typed_tensor);
+  } else {
+    // at this point we know dst is an XLA tensor
```

Comment:
Could you help me understand why we know dst is an XLA tensor here? I guess I am trying to understand the use case of this function. What does checking `self_tensor` means? Are we checking if they are empty tensor or CPU tensor?

Reply:
I wrote this function to be similar to `AtenXlaType::_copy_from`, in that it handles all 3 cases:
* `self` is on CPU and `dst` is on XLA
* `self` is on XLA and `dst` is on CPU
* they're both on XLA
The main difference is just that in the case that `dst` is on XLA, it handles resizing properly. I got the resizing logic mostly by looking at the function that the codegen [used to call](https://github.com/pytorch/xla/blob/7b538aacc9c9620b7533e89ccfbdbfef8e46fad1/torch_xla/csrc/aten_xla_bridge.cpp#L159).

> Could you help me understand why we know dst is an XLA tensor here? 

In the `} else {` section here, we know that both `bridge::TryGetXlaTensor(self)` and `bridge::TryGetXlaTensor(dst)` returned valid `XLATensor` objects, which I took to mean that they both live on XLA. So this piece of the code handles the cases where both `self` and `dst` live on XLA, and I took the corresponding logic from [here](https://github.com/pytorch/xla/blob/7b538aacc9c9620b7533e89ccfbdbfef8e46fad1/torch_xla/csrc/aten_xla_bridge.cpp#L173) to figure out how to copy and resize the dst XLA tensor.

This piece of logic gets hit inside of the generated `out` wrappers. I pasted an example codegen'd `out` wrapper for `add` below:
```
at::Tensor & wrapper_out_add_out(const at::Tensor & self, const at::Tensor & other, co
  XLA_FN_TRACK(3);
  TF_VLOG(3) << "XLA wrapper_out_add_out :" << " self=" << self.toString() << " other=

  auto wrapper_out_add_out_tmp = wrapper_Tensor_add(self, other, alpha);
  at::_copy_from_and_resize(wrapper_out_add_out_tmp, out);
  return out;
}
```